src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.cc

```diff
@@ -249,21 +249,39 @@ static tsi_result alts_zero_copy_grpc_protector_max_frame_size(
   return TSI_OK;
 }
 
+static tsi_result alts_zero_copy_grpc_protector_read_frame_size(
+    tsi_zero_copy_grpc_protector* self, grpc_slice_buffer* protected_slices,
+    uint32_t* frame_size) {
+  if (self == nullptr || frame_size == nullptr) return TSI_INVALID_ARGUMENT;
+  alts_zero_copy_grpc_protector* protector =
+      reinterpret_cast<alts_zero_copy_grpc_protector*>(self);
+
+  if (protector->parsed_frame_size == 0) {
+    // We have not parsed frame size yet. Parses frame size.
+    if (!read_frame_size(protected_slices, &protector->parsed_frame_size)) {
+      return TSI_DATA_CORRUPTED;
```

Comment:
I'm not sure this is what we want to do: if `protected_slices` contains only the first half of the header, then we are going to return data corrupted, which will cause the connection to get closed.